packages/astroport/src/restricted_vector.rs

```diff
@@ -0,0 +1,92 @@
+use crate::DecimalCheckedOps;
+use cosmwasm_std::{Decimal, StdError, StdResult, Uint128};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::fmt::Display;
+
+/// Vec wrapper for internal use.
+/// Some business logic relies on an order of this vector, thus it is forbidden to sort it
+/// or remove elements. New values can be added using .update() ONLY.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct RestrictedVector<K, V>(Vec<(K, V)>);
+
+pub trait Increaseable
+where
+    Self: Sized,
+{
+    fn increase(self, new: Self) -> StdResult<Self>;
+}
+
+impl<K, V> RestrictedVector<K, V>
+where
+    K: PartialEq + Display,
+    V: Copy + Increaseable,
+{
+    pub fn new(key: K, value: V) -> Self {
+        Self(vec![(key, value)])
+    }
+
+    pub fn get_last(&self, key: &K) -> StdResult<V> {
+        self.0
+            .last()
+            .filter(|(k, _)| k == key)
+            .map(|(_, v)| v)
+            .cloned()
+            .ok_or_else(|| StdError::generic_err(format!("Key {} not found", key)))
+    }
+
+    pub fn update(&mut self, key: K, value: V) -> StdResult<V> {
```

Comment:
I think we may just borrow 'key' here like this ```..., key: &K, ... ```   
Thus we can escape cloning of 'key' every time we call this function.

Reply:
Ok, done